Review (#1)
geun-son/armeria: site/src/pages/tutorials/thrift/blog/create-server.mdx

---
menuTitle: "Create a server"
order: 1
category: thrift
type: step
targetLang: java
---

# Creating a server

> As the first step of the tutorial, we'll write a Thrift server that serve a `hello` service. 

Reviewer: ```suggestion
As the first step of the tutorial, we'll write a Thrift server with a dummy service.
```

<Tip>

> 

Reviewer: `<TutorialSteps current={1} />`

여기 TutorialSteps UI가 없어요.

> Please refer to [Running a Thrift service](/docs/server-thrift) for creating Thrift server.

Reviewer: ```suggestion
Please refer to [Running a Thrift service](/docs/server-thrift) for creating a Thrift server.
```

</Tip>

> ## 1. Create a Thrift file

Reviewer: Thrift file > thrift file (또는 .thrift file)

'.thrift 확장자를 가진 파일 형식'을 얘기할 때는 소문자 표기가 나을 것 같습니다. 표기 통일하면 좋을 것 같아요.

[Thrift 문서](https://thrift.apache.org/)도 좀 혼용하고 있긴 합니다만..
<img width="672" alt="스크린샷 2023-03-22 오전 2 03 57" src="https://user-images.githubusercontent.com/23310187/226686574-e1417027-a8d7-4780-b03d-9311d62b62b2.png">

> Create a file, `blog.thrift` inside the directory, `{project_root}/src/main/thrift` each server and client.

Reviewer: 제가 처음 튜토리얼을 보고 따라하면, 시작하자마자 여기서 막힐 것 같습니다. '알려주는 디렉토리 위치는 하난데 어떻게 같은 이름의 파일을 두 개 만들라는 거지?'라는 의문이 생깁니다.

**제안 1:** 
프로젝트를 두 개 만들어야 한다는 정보를 index.mdx에서 미리 알려주면 어떨까요? REST 튜토리얼처럼 Sample service structure를 주는 것도 좋을 것 같아요.

**제안 2:**
서버/클라이언트 프로젝트 각각 따로 만들지 말고, 하나로 만드는 건 어떨까요? 제가 gRPC 튜토리얼 따라 할때는 최종적으로 그런 형태가 되었었거든요. 이렇게 되면 Thrift 파일 하나를 서버/클라이언트에서 공유할 수 있습니다. 현실에서는 프로젝트를 하나로 만들진 않겠지만, 프로젝트를 각각 만들어서 두 개 왔다갔다 하느라 헷갈리는 것 보다 이게 나을수도 있겠다 싶어서요. 아래는 트리 구조 예시입니다.
```
* src/
    * main/
        * java/
            * example.armeria/
                * server.blog.thrift/
                    * BlogServiceExceptionHandler
                    * BlogServiceImpl
                    * Main
                * client.blog.thrift/
                    * BlogClient.java
        * thrift/
            * blog.thrift
    * test/
        * (생략)
* build.gradle
```
저는 개인적으로 제안 2가 제일 나은 것 같아요.

**제안 3:**
클라이언트를 구현해서 호출하는 대신, 테스트 코드를 실행해서 구현을 검증하도록 유도하는 방법도 괜찮을 것 같아요. 테스트 코드가 샘플 코드로 이미 있기도 하고요. 다만 문서 수정사항은 더 커지겠네요.

> Create a file, `blog.thrift` inside the directory, `{project_root}/src/main/thrift` each server and client.

Reviewer: ```suggestion
Create two `blog.thrift` files inside the directory `{project_root}/src/main/thrift` for server and client.
```
만약에 **제안1**로 한다면 참고해주세요.

}
```

## 3. Compile the thrift file

Compile the `blog.thrift` file to generate Java code. 

```bash
./gradlew build
```

<Tip>

  Please refer to [build.gradle](https://github.com/line/armeria-examples/blob/master/tutorials/thrift/build.gradle) for generating code with [Gradle Thrift Plugin](https://github.com/jruyi/thrift-gradle-plugin).

</Tip>

Find the Java code output in the directory, `{project_root}/build/generated-sources/thrift/gen-java/example/armeria/blog/thrift/`.

## 4. Declare the blog service and override the `hello` method 

Create a file `BlogServiceImpl.java`. Declare the service and override the `hello` method that we've defined in the blog.thrift file. 

```java filename=BlogServiceImpl.java
package example.armeria.server.blog.thrift;

import example.armeria.blog.thrift.BlogService;

import org.apache.thrift.TException;
import org.apache.thrift.async.AsyncMethodCallback;

public class BlogServiceImpl implements BlogService.AsyncIface{
    @Override
    public void hello(String request, AsyncMethodCallback<String> resultHandler) {
        resultHandler.onComplete("Hello, " + request + '!');
    } 
}
```

## 5. Add a service to a server

> Build a service and server using Armeria's <type://ServerBuilder> to serve our service.

Reviewer: ```suggestion
Build a server and add service to the server using Armeria's <type://ServerBuilder>.
```

  ```java filename=Main.java
  package example.armeria.server.blog.thrift;

  import org.slf4j.Logger;
  import org.slf4j.LoggerFactory;

  public final class Main {
    private static final Logger logger = LoggerFactory.getLogger(Main.class);
  }
  ```
2. Create a service instance using Armeria's <type://THttpService#builder()>.
  ```java filename=Main.java
  import com.linecorp.armeria.server.thrift.THttpService;

  public final class Main {
    private static Server newServer(int port) throws Exception {
        final THttpService tHttpService =
                THttpService.builder()
                            .addService(new BlogServiceImpl())
                            .build();
    }
  }
  ```
3. Build and return a new server instance using Armeria's <type://ServerBuilder>.
  ```java filename=Main.java
    public final class Main {
      private static Server newServer(int port) throws Exception {
        ...
        return Server.builder()
                     .http(port)
                     .service("/thrift", tHttpService)
                     .build();
      }
    }
  ```

> ## 6. Serve the service

Reviewer: Serve the service > Start the server (혹은 Run the server)

더 적절한 의미의 표현 제안합니다.

Create a server instance in the `main()` method.

```java filename=Main.java
  public static void main(String[] args) throws Exception {
      final Server server = newServer(8080);

      server.closeOnJvmShutdown().thenRun(() -> {
          logger.info("Server has been stopped.");
      });

      server.start().join();
  }
  ```

## 7. Test run

Start the server by running the `Main.main()` method on your IDE or using Gradle.

```bash
./gradlew run
```

Your server is running if you see the following message.

```bash
[armeria-boss-http-*:8080] INFO com.linecorp.armeria.server.Server - Serving HTTP at /[0:0:0:0:0:0:0:0]:8080 - http://127.0.0.1:8080/
```

## What's next

> In this step, we've defined a thrift file for our service and generated Java code. And then, we've created and added a Thrift service to a server. 

Reviewer: See [Running a Thrift service](/docs/server-thrift) for more information.

[Running a Thrift service](/docs/server-thrift) 문서 링크를 같은 페이지 위쪽에서 팁으로 제공하는 대신에, 여기에 이렇게 덧붙이면 어떨까요? 튜토리얼을 시작하는 시점 보다는, 얼추 따라한 뒤에야 메뉴얼이 궁금할 것 같아서요.
